### DISCUSSION TITLE
Subject: Property 'next' does not exist on type 'Observable<any>'.

### DISCUSSION THREAD

**Discussion by hoclun-rigsep** [ReactiveX/rxjs#6440] [Q&A]
```js
import { Subject } from 'rxjs';
import { tap } from 'rxjs/operators';

  let a = new Subject().pipe(tap(console.log));
  a.subscribe(console.log);
  a.next("Hello World");

```

This code runs as I expect but I get the following type error:

> Property 'next' does not exist on type 'Observable<any>'.

I can't imagine I'm the first person to notice this, but how should I take it?
It seems to me that `new Subject().pipe(...)` should return some type
that's like Observable but with `.next()`, or that `.next()` should be private
here (but why?).

**Comment by raymonddavis** [comment]
This is what I think is happening. Hope it helps.

The subject has next defined. https://github.com/ReactiveX/rxjs/blob/f47425d349475231c0f3542bb6ecef16a63e933a/src/internal/Subject.ts#L56

The observable does not it, it works because the subject extends observable so you can use pipe which from a typing point of view has no knowledge of the subject type. Basically, the next prop is just being left on the observable after the pipe changes its type into an observable.

I personally separate my input streams from my output streams, like read and write segregation. 
Subject is an input stream in my option I kept it separate so I can push new items into it.
Then any time I change it or add operators onto it becomes a new output stream.
This keeps the typing happy and helps clearly show where inputs and outputs come from.
```ts
let input = new Subject();
let output = input.pipe(tap(console.log))
let output2 = output.pipe(tap(console.log))

output2.subscribe(console.log);

input.next('Hello World');
```

**Reply by hoclun-rigsep** [reply]
Thanks for the discussion, @raymonddavis 

> Basically, the next prop is just being left on the observable after the pipe changes its type into an observable.

Would you agree that either the method should be made unavailable or the type should accurately reflect its presence?

```
const _Bucket = new BehaviorSubject(0)
const Bucket = _Bucket.pipe(scan((a,b) => a+b));

Bucket.subscribe(console.log);

Bucket.next(1); //perhaps in a click handler
Bucket.next(1);
Bucket.next(1);
Bucket.next(1);
Bucket.next(1);
console.log(Bucket.getValue());
```

I find this sort of thing fairly intuitive, and it works exactly as expected! Except for the compiler errors, and except, that it seems that while `.next()` is available on the result of the pipe, `.getValue()` isn't. I wonder if this is intentional, and how it ought to be. When I have time I'll read the source for how exactly the resulting Observable inherits what it inherits.

**Reply by raymonddavis** [reply]
**_This is what I think is happening_**

BehaviorSubject extends Subject
Subject extends Observable
 
Pipe returns Observable.
So the typing will not go "up 2 levels" into BehaviorSubject or into Subject.

The reason that getValue() throws an error is because its not a prop on the returned value. (javascript error)

The reason next() is just a typing error I believe is because without next, error, and complete being exposed how would things be able to connect/chain to each other.

I think method should be unavailable but I honestly just follow the typing.
If you start calling functions that are not in the typing it will just cause confusion in the future for other developers within your team. Also stuff that is not in the typing could change between versions without your knowledge.

I don't know if anything could be done about this. I don't think it is an issue but that is my option.

**Reply by hoclun-rigsep** [reply]
@raymonddavis Okay! I think I agree with almost everything you say there. I also agree with the idea of splitting the Subject up—but it's kind of interesting, Subject glues these things together and then we pull them apart, see eg <https://react-rxjs.org/docs/api/utils/createSignal>

**Reply by raymonddavis** [reply]
@hoclun-rigsep I just realized If you add a `.asObservable()` onto the subject before you pipe next() will throw an error.
